oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java

```diff
@@ -0,0 +1,813 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+
+import org.apache.jackrabbit.oak.api.CommitFailedException;
+import org.apache.jackrabbit.oak.plugins.document.memory.MemoryDocumentStore;
+import org.apache.jackrabbit.oak.plugins.document.util.LeaseCheckDocumentStoreWrapper;
+import org.apache.jackrabbit.oak.spi.commit.CommitInfo;
+import org.apache.jackrabbit.oak.spi.commit.EmptyHook;
+import org.apache.jackrabbit.oak.spi.state.NodeBuilder;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.stats.Clock;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import junitx.util.PrivateAccessor;
+
+/**
+ * Tests unrecovered revisions on a document
+ */
+public class UnrecoveredRevisionTest {
+    
+    interface BreakpointCallback {
+        
+        void breakpointCallback(String contextMsg);
+        
+    }
+    
+    class ControllableMemoryDocumentStore extends MemoryDocumentStore {
+        
+        private final AtomicInteger remainingOps;
+        private BreakpointCallback callback;
+        private Set<Collection<?>> restrictedCollections = new HashSet<>();
+        private List<UpdateOp> writtenUpdateOps = new LinkedList<>();
+        
+        ControllableMemoryDocumentStore(int remainingOps, BreakpointCallback callback) {
+            this.remainingOps = new AtomicInteger(remainingOps);
+            this.callback = callback;
+            // restrict only NODES by default
+            this.restrictedCollections.add(Collection.NODES);
+        }
+        
+        public void clearOps() {
+            writtenUpdateOps.clear();
+        }
+
+        public void printOps() {
+            for (UpdateOp updateOp : writtenUpdateOps) {
+                System.out.println("written updateOp: " + updateOp.toString());
+            }
+        }
+        
+        private void registerWrittenUpdateOp(Collection<?> collection, UpdateOp updateOp) {
+            if (restrictedCollections.contains(collection)) {
+                writtenUpdateOps.add(updateOp);
+            }
+        }
+
+        private void registerWrittenUpdateOp(Collection<?> collection, List<UpdateOp> trimmedOps) {
+            if (restrictedCollections.contains(collection)) {
+                writtenUpdateOps.addAll(trimmedOps);
+            }
+        }
+
+        private void setBreakpointCallback(BreakpointCallback callback) {
+            this.callback = callback;
+        }
+        
+        @SuppressWarnings("unused")
+        private void setRestrictedCollections(Collection<?>... collections) {
```

Comment:
Can we remove this un-used method or do we plan to use it in the future?

Reply:
I'd prefer to have it in for now? I was expecting (or let's say hoping) it to eventually be used, once we go more fancy with test coverage.